Check arity consistency when constructing `SqlFunctionExpression`s

ranma42 opened the pull request:
It is currently possible to construct `SqlFunctionExpression`s in which the size of `arguments` does not match the size of `argumentsPropagateNullability`.

While this is definitely a problem of mis-using the API, it is both easy to trip in it (in fact there is an occurrence of such problem in the repo 😅 ) and easy for EFCore to provide better feedback when this happens.

ranma42 commented:
I believe that the commit fixing the translation of `random()` in sqlite is likely uncontroversial, while the one adding checks probably needs some discussion:
 - is this considered a breaking change? (previously it was possible to create more `SqlFunctionExpression`s; EFCore would have been unable to handle them properly, but someone might have been creating them with the inconsistency and later on transforming them with another library before passing them to EFCore)
 - I used `ArgumentException` as that seemed the most appropriate for this case; for a similar problem `SqlUnaryExpression` and `SqlBinaryExpression` use `InvalidOperationException`, which is meant to be "thrown when a method call is invalid for the object's current state" [source](https://learn.microsoft.com/en-us/dotnet/api/system.invalidoperationexception?view=net-8.0); is it better to use `InvalidOperationException` for this case as well?
 - should we do this check at all?

maumar commented:
It is a breaking change. While we already rely on the assumption that arguments and propagates nullability information tables are in sync, those were only being used when processing IS NULL/IS NOT NULL expressions. With this change (and nullability computation work in https://github.com/dotnet/efcore/pull/33814) more people get affected. However, I agree with @cincuranet that the impact is relatively small.

One thing to note, argument propagates nullability is somewhat complicated concept - some users may not fully understand it and as such leave the array empty (and hey, it happened to have worked! ;) ) - Perhaps it's beneficial for the exception message to provide some guidance - as in, if not sure use "false", or (in the future) link to a documentation section, once we have it.

maumar commented:
wrt exception type, I also like InvalidOperationException better, EF tends to favor it for almost all errors we issue.

ranma42 commented:
I switched the exception type to `InvalidOperationException` and moved the strings to `RelationalStrings` for consistency with `Sql{Bi,U}naryExpression`.

cincuranet commented:
Looks like we have 2 (strongly related :)) failures:
```
Microsoft.EntityFrameworkCore.Query.SpatialQuerySqliteTest.GeometryType(async: ...)
System.InvalidOperationException : 'rtrim' was constructed with 2 arguments, but the nullability was defined for 1 arguments.
```
```
Microsoft.EntityFrameworkCore.Query.SpatialQuerySqliteTest.OgcGeometryType(async: ...)
System.InvalidOperationException : 'rtrim' was constructed with 2 arguments, but the nullability was defined for 1 arguments.
```

ranma42 commented:
ouch, sorry, I did not notice that locally I was not running with `SpatiaLite`.
I am running on linux so unfortunately it requires customizing the csprojs (as explained here https://learn.microsoft.com/en-us/ef/core/providers/sqlite/spatial ).
I enabled it for some runs, but I probably forgot to apply those changes while rewriting the baselines in this case.
I will fix ASAP 😇

ranma42 commented:
side note: in addition to the arity mismatches there seem to be some inconsistency in the nullability configuration in the codebase.
I might eventually be able to check that as well, but I believe that can be done in a separate PR.

cincuranet commented:
> but I believe that can be done in a separate PR

Agreed.

cincuranet commented:
Thanks for the contribution.

ranma42 commented:
thank you for your patience and for guiding me :)
